portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -14,32 +14,94 @@
   @Override
-  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    ArrayList<String> messages = new ArrayList<String>();
-    messages.add("i am writing this again");
-    messages.add("because I accidentally deleted it once");
-    messages.add("whoops");
-    String json = convertToJson(messages);
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+      String title = getParameter(request,"title", "");
+      String body = getParameter(request,"comment", "");
+      long timestamp = System.currentTimeMillis();
 
-    response.setContentType("application/json;");
-    response.getWriter().println(json);
+      Entity commentEntity = new Entity("comment");
+      commentEntity.setProperty("title",title);
+      commentEntity.setProperty("body", body);
+      commentEntity.setProperty("timestamp",timestamp);
+
+      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+      datastore.put(commentEntity);
+      response.sendRedirect("/feedback.html");
   }
-  private String convertToJson(ArrayList<String> messages){
-      Gson gson = new Gson();
-      String json = gson.toJson(messages);
-      return json;
+
+  private String getParameter(HttpServletRequest request, String name, String defaultValue) {
+    String value = request.getParameter(name);
+    if (value == null) {
+      return defaultValue;
+    }
+    return value;
+  }
+   public class Comment{
+      String title;
+      String body;
+      long timestamp;
+      long id;
+      public Comment(String title, String body,long timestamp, long id){
+          this.title = title;
+          this.body = body;
+          this.timestamp = timestamp;
+          this.id = id;
+      }
+      public String getTitle(){
+          return title;
+      }
+      public String getBody(){
+          return body;
+      }
+      public long getTimestamp(){
+          return timestamp;
+      }
+      public long getId(){
+          return id;
+      }
   }
+
+@Override
+public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    Query query = new Query("comment").addSort("timestamp", SortDirection.DESCENDING);
+    
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    ArrayList<Comment> comments = new ArrayList<>();
+    for (Entity entity : results.asIterable()) {
+        String title = (String) entity.getProperty("title");
+        String body = (String) entity.getProperty("body");
+        long timestamp = (long) entity.getProperty("timestamp");
+        long id = entity.getKey().getId();
+
+        Comment comment = new Comment(title, body, timestamp, id);
+        comments.add(comment);
+    }
+    Gson gson = new Gson();
+
+    response.setContentType("application/json;");
+    response.getWriter().println(gson.toJson(comments));
```

Comment:
The whole process of converting from datastore Entity to JSON here seems like magic. It's not obvious to me, as a reader, how the resulting JSON will look.

Normally I'd ask for a unit test to make sure the JSON looks how we expect. But the STEP program hasn't covered how to write those yet, have they?

Also, normally I'd say let's simplify this and produce the JSON without using Gson, since AFAICT Gson is designed for something we're not doing: turning Java objects into strings and then back into Java objects. But I think the STEP team asked you to use Gson?

Reply:
No, the tutorial has not covered testing yet. I'm not sure if that's something I have not gotten to yet in the tutorial, or if that's going to be taught a different week. Should I look into creating unit tests?
Yeah the STEP tutorial has you using Gson for everything. I'm not sure why.

Reply:
Test driven development will be covered in week 5 of your internship (Project 4) so I wouldn't worry too much about it now, but feel free to start doing some preliminary research if you have time!